File: leet-code/49. Group Anagrams/49_group_anagrams.py
from collections import defaultdict


def group_anagrams(strs: list[str]) -> list[list[str]]:
    dic = defaultdict(list)

    for idx, s in enumerate(strs):
        sorted_s = ''.join(sorted(s))
        dic[sorted_s].append(s)

    return list(dic.values())


print(group_anagrams(["eat","tea","tan","ate","nat","bat"]))

File: leet-code/49. Group Anagrams/49_group_anagrams.dart
List<List<String>> groupAnagrams(List<String> strs) {
  Map<String, List<String>> anagramMaps = {};
  
  for (int i = 0; i < strs.length; i++) {
    List<String> s = strs[i].split('');
    s.sort();
    String sortedS = s.join('');
    
    if (anagramMaps.containsKey(sortedS)) {
      anagramMaps[sortedS]!.add(strs[i]);
    } else {
      anagramMaps[sortedS] = [strs[i]];
    }
  }

  List<List<String>> answer = [];
  anagramMaps.forEach((key, value) { answer.add(value); });

  return answer;
}

void main() {
  print(groupAnagrams(["eat","tea","tan","ate","nat","bat"]));
}